### ISSUE TITLE
Keep clocks around longer

### ISSUE THREAD

**Issue by adriancole** [openzipkin/brave#656]
Processor spans happen after consumer ones.. the current "clock sharing" is scoped to span objects. The problem is that a consumer span opens and closes before a processor one does. When not using JRE 9, this can create a small amount of distracting skew

<img width="965" alt="screen shot 2018-03-13 at 3 17 39 pm" src="https://user-images.githubusercontent.com/64215/37327546-b5a852d6-26d1-11e8-89a2-51f23c4780fe.png">

In rabbit, we can fake this out as we are simulating spans anyway https://github.com/openzipkin/brave/pull/653

In general, I think we need to change the implicit clock sharing to be based on the trace, not the parent. Also, it should try to keep clocks around a while in case there's some small lag before the next span starts.

**Comment by Logic-32** [comment]
Looked at your commit @adriancole, was the bug here that you finished a parent span prior to initializing the child span?  In doing so, the SpanMap got cleared out so the Clock was no longer associated with the TraceContext?

I ask because, associating a clock with a TraceContext should mean that clock sharing is based on the trace.  But I can see how storing it in the SpanMap would make it not actually behave that way.

**Comment by adriancole** [comment]
yep you got it. To do this for gaps (or even concurrent inbound requests
for the same trace), we'd need a different cache, based on traceId I think.
